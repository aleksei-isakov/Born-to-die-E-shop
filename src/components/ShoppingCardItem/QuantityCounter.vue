<template>
  <div class="quantity__wrapper">
    <BaseTextFilledButton>-</BaseTextFilledButton>
    <input readonly :value="quantity" class="quantity-field" />
    <BaseTextFilledButton>+</BaseTextFilledButton>
  </div>
</template>

<script>
import { BaseTextFilledButton } from '@/base_components';

export default {
  name: 'QuantityCounter',

  components: { BaseTextFilledButton },

  props: {
    quantity: {
      type: Number,
      require: true,
      default: 1
    }
  }
};
</script>

<style lang="scss" scoped>
@import '@/scss/CustomVariables.scss';

.quantity__wrapper {
  width: 30%;
  display: inline-flex;
  flex-grow: 0;
  justify-content: flex-end;
}

.quantity-field {
  width: 60px;
  text-align: center;
  border: 2px solid #e4e4e4ff;
}

@media screen and (max-width: $tablet-size) {
  .quantity__wrapper {
    width: auto;
  }

  .quantity-field {
    width: 30px;
  }
}
</style>
