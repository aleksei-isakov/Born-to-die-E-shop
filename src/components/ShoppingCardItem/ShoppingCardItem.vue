<template>
  <div class="shoppingCard__wrapper">
    <img :src="getImage" class="cart-item-img" />
    <div class="shopping-cart-item-description">{{ description }}</div>
    <div class="product-price">{{ getPrice }} $</div>
    <quantity-counter :quantity="quantity" />
    <shopping-cart-trash-icon @delete-item="deleteFromCart(id)" />
  </div>
</template>

<script>
import defaultImage from '@/assets/defaultImage.jpg';
import QuantityCounter from '@/components/ShoppingCardItem/QuantityCounter';
import ShoppingCartTrashIcon from '@/components/ShoppingCartTrashIcon/ShoppingCartTrashIcon.vue';
import { mapActions } from 'vuex';

export default {
  name: 'ShoppingCardItem',

  components: { QuantityCounter, ShoppingCartTrashIcon },

  props: {
    image: {
      type: String,
      require: true,
      default: defaultImage
    },

    index: {
      type: Number,
      require: true,
      default: 0
    },

    id: {
      type: String,
      require: true,
      default: ''
    },

    description: {
      type: String,
      require: true,
      default: ''
    },

    quantity: {
      type: Number,
      require: true,
      default: 1
    },

    price: {
      type: Number,
      require: true,
      default: 0
    }
  },

  computed: {
    getImage() {
      return this.image ? this.image : defaultImage;
    },

    getPrice() {
      return this.price * this.quantity;
    }
  },
  methods: {
    ...mapActions('ShoppingCartModule', ['deleteFromCart'])
  }
};
</script>

<style lang="scss" scoped>
@import '@/scss/CustomVariables.scss';

.cart-item-img {
  align-content: flex-start;
}

.shopping-cart-item-description {
  margin-right: 10px;
  max-width: 300px;
}

.shoppingCard__wrapper {
  width: 80%;
  display: inline-flex;
  color: $font-color-text;
  max-width: 800px;
  align-items: center;
  flex-grow: 1;
  justify-content: space-between;
  border: 1px solid #e4e4e4ff;
  box-shadow: 0 3px 2px -2px gray;
  padding: 10px;
}

.product-price {
  font-size: 1.5rem;
  color: $font-color-subtitle;
}

.delete-icon {
  cursor: pointer;
}

@media screen and (max-width: $tablet-size) {
  .shoppingCard__wrapper {
    width: 99%;
    padding: 10px;
  }

  .product-price {
    font-size: 1.3rem;
    color: $font-color-subtitle;
  }

  .cart-item-img {
    max-width: 100px;
  }
}
</style>
