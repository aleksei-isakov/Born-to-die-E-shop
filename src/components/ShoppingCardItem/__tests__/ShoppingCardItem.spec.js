import Vue from 'vue';
import Vuetify from 'vuetify';
import { mount } from '@vue/test-utils';
import ShoppingCardItem from '@/components/ShoppingCardItem/ShoppingCardItem';
import defaultImage from '@/assets/defaultImage.jpg';
Vue.use(Vuetify);

describe('ShoppingCardItem', () => {
  let wrapper;
  const name = 'name';
  const price = 35;
  const quantity = 1;
  const id = '0';
  let vuetify = new Vuetify();
  const imageList = [
    'http://placeimg.com/640/480',
    'http://placeimg.com/640/480',
    'http://placeimg.com/640/480'
  ];

  beforeEach(() => {
    wrapper = mount(ShoppingCardItem, {
      vuetify,
      propsData: {
        images: imageList,
        id: id,
        name: name,
        quantity: quantity,
        price: price
      }
    });
  });

  afterEach(() => {
    wrapper.destroy();
  });

  test('is a Vue instance', () => {
    expect(wrapper.isVueInstance).toBeTruthy();
  });

  test('should return correct props', () => {
    expect(wrapper.props().name).toBe(name);
    expect(wrapper.props().price).toBe(price);
    expect(wrapper.props().quantity).toBe(quantity);
    expect(wrapper.props().images).toBe(imageList);
    expect(wrapper.props().id).toBe(id);
  });

  test('Should take the first image if images array is not empty', () => {
    const localThis = { images: imageList };

    expect(ShoppingCardItem.computed.getImage.call(localThis)).toBe(
      imageList[0]
    );
  });

  test('Should get default image when no images in the props', () => {
    const localThis = { images: null };

    expect(ShoppingCardItem.computed.getImage.call(localThis)).toBe(
      defaultImage
    );
  });

  test('match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
