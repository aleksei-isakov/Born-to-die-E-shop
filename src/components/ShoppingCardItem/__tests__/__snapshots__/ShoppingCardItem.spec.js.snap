// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShoppingCardItem match snapshot 1`] = `
<div class="shoppingCard__wrapper"><img src="" class="cart-item-img">
  <div class="shopping-cart-item-description">description</div>
  <div class="product-price">35 $</div>
  <div class="quantity__wrapper"><button class="base-button base-text-button base-text-filled-button custom-blue">-</button> <input readonly="readonly" class="quantity-field"> <button class="base-button base-text-button base-text-filled-button custom-blue">+</button></div>
  <div class="trash-icon__wrapper">
    <div class="title-icon trash-icon"><img src="" width="25" alt="icon"></div>
  </div>
</div>
`;
